File: graphr/src/graphr/algorithms/ConnectionDistanceAgent.java
package graphr.algorithms;

import java.util.ArrayList;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import graphr.data.PrimData;
import graphr.processing.Agent;
import graphr.processing.VertexProcessingFacade;

public class ConnectionDistanceAgent extends Agent {
	
	private static Logger log = LogManager.getLogger(); 
	
	int sourceId;
	int distance;
	int maxDistance;

	//-- Implementations of abstract stuff
	
	public ConnectionDistanceAgent(int sourceId, int maxDistance) {
		this.sourceId = sourceId;
		this.maxDistance = maxDistance;
	}
	
	public ConnectionDistanceAgent getCopy() {
		ConnectionDistanceAgent a = new ConnectionDistanceAgent(sourceId, maxDistance);
		a.distance = distance;
		a.v = v;
		return a;
	}

	@Override
	public void runStep() {
		log.debug("vid: " + v.getId());
		
		if ((sourceId == v.getId()) && (v.getValue("distance") == null)){
			modifyDistanceAndBroadcast(0);
		} else {
			PrimData vDistance = v.getValue("distance");
			log.debug("distance: " + (vDistance == null ? "null" : vDistance.i()));
			if ((vDistance == null) || (vDistance.i() > this.distance)) {
				modifyDistanceAndBroadcast(this.distance);
			} 
		}
	}
	
	//-- Elements of algorithm
	
	public void modifyDistanceAndBroadcast(int distance) {
		v.setValue("distance", new PrimData(distance));
		v.setValue("vislabel", new PrimData(getVisLabel()));
		this.distance = distance + 1;
		v.broadcast();
	}
	
	public String getVisLabel() {
		return v.getValue("label").s() + ": " + v.getValue("distance").i(); 
	}

}

File: graphrs/graphrs/algorithms/ConnectionDistanceAgent.swift
//
//  ConnectionDistanceAgent.swift
//  graphrs
//
//  Created by Marcus Sellmann on 30.03.16.
//  Copyright © 2016 CITBDA. All rights reserved.
//

class ConnectionDistanceAgent : Agent {
    var v: VertexProcessingFacade
    var sourceId: UInt64
    var distance: Int = 0
    var maxDistance: Int
    
    // -- Implementations of abstract stuff
    init(sourceId: UInt64, maxDistance: Int, vertexProcessingFacade: VertexProcessingFacade) {
        self.sourceId = sourceId
        self.maxDistance = maxDistance
        self.v = vertexProcessingFacade
    }
    
    func getCopy() -> Agent {
        let a = ConnectionDistanceAgent(sourceId: sourceId, maxDistance: maxDistance, vertexProcessingFacade: v)
        a.distance = distance
        
        return a;
    }
    
    func runBefore() {
    }
    
    func runStep() {
        if ((sourceId == v.getId()) && (v.getValue("distance") == nil)) {
            modifyDistanceAndBroadcast(0);
        } else {
            do {
                let vDistance = try (v.getValue("distance")!.i())
                
                if (vDistance > self.distance) {
                    modifyDistanceAndBroadcast(self.distance)
                }
            } catch {}
        }
    }
    
    func runAfter() {
    }
    
    // -- Elements of algorithm
    
    func modifyDistanceAndBroadcast(distance: Int) {
        v.setValue("distance", value: PrimitiveData(i: distance))
        v.setValue("vislabel", value: PrimitiveData(s: getVisLabel()))
        self.distance += 1
        v.broadcast()
    }
    
    func getVisLabel() -> String {
        var d = 0
        var cs = ""
        
        do {
            d = try v.getValue("distance")!.i()
        } catch {}
        
        switch (d) {
            case 0:
                cs = ""
                break;
            case 1:
                cs = ": directly connected"
                break;
            case 2:
                cs = ": 2nd degree"
                break;
            case 3:
                cs = ": 3rd degree"
                break;
            default:
                cs = ": \(d)th degree"
        }
        
        return String(v.getValue("label")) + cs
    }
    
    func setUsedEdge(e: Edge) {
    }
}